rfcs/2021-05-05-7351-framing-and-codecs.md

```diff
@@ -0,0 +1,183 @@
+# RFC 7351 - 2021-05-05 - Framing and Codecs
+
+This RFC discusses a standardized way of specifying framing and codecs across
+sources and sinks. _Framing_ is concerned with turning sequences of bytes into
+byte frames (which indicate the boundaries of a complete message). Codecs
+consist of _decoders_, which deserialize a byte frame into structured data, and
+_encoders_, which serialize structured data into a byte frame.
+
+Conceptually, we want reusable pieces of logic which would allow us to collapse
+a source + (decoder) transform into a source with decoder config, and to
+collapse an (encoder) transform + sink into a sink with encoder config, in a way
+that is transparent to the user. These transforms merely convert between
+encoding formats, e.g. from `bytes` to `json` and vice versa.
+
+## Scope
+
+The scope of this RFC concern at which level framing and decoding operates, how
+these framers and codecs can be configured, and how they can be shared in a
+uniform way. It does not cover any specific implementation for framing or a
+codec. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and codecs, such that each source and sink may implement them in their
+own way. This can introduce inconsistencies and additional maintenance burden.
+Components have their own defaults, making behavior unpredictable and surprising
+users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+Since decoding and encoding should apply universally to all sources and sinks,
+it should be configurable by a shared field. This is accomplished by the
+`SourceOuter` and `SinkOuter` wrappers:
+
+```rust
+// Newly introduced wrapper, analogously to `SinkOuter`.
+struct SourceOuter {
+    framing: Framer,
```

Comment:
Do we want to include a `framer` for all sources, even those which already use message-oriented protocols (e.g. kafka)?

Reply:
That is a good question, it will indeed only apply to a relatively small number of sources.

The current implementation expects a collection of `FramingConfig`s in that place and defaults to an empty array. The source implementations which do not use framing can simply ignore the `framing` field from the `SourceContext`.

Reply:
Ignoring `framing` field could be surprising for users. The best option would be to deny parsing such configs, or at least we can report a warning.